### PR TITLE
Replace required pipeline with final pipeline

### LINKED ISSUE #49247: Improve flexibility over index.required_pipeline
**Describe the feature**:

Following up on https://github.com/elastic/elasticsearch/pull/46847 there are a couple of cases where we want to ensure that a specific pipeline is run on any documents that are ingested into an index. For example, you may want to set the `event.ingested` timestamp or ensure that the name of the API Key used is present in the document.

At the same time, we want to give users the flexibility they currently have to use a pipeline of their choosing to process the incoming data. We have `index.required_pipeline`, but it doesn't come with the flexibility we'd like.

@skearns64 suggested:
> Sounds like we need an "append" pipeline, or an option to required to be "run first or run last"

If "append pipeline" means that Elasticsearch will automatically run the "append pipeline" on every indexed document after the pipeline specified with the request has been run, it sounds like the "append pipeline" option would solve the use-cases I'm familiar with.

I've not heard a compelling use case for "run first", but they could exist.

some questions that come to mind:
- Does "append pipeline" let users specify a list of pipelines to append or only a single pipeline. You can achieve the same functionality by combining pipelines, but I can imagine it would be convenient to be able to specify a list.
- how will it work with `index.default_pipeline` and `index.required_pipeline`
- I don't know that `index.required_pipeline` has any use case that `index.append_pipeline` does not solve, but that could be due to lack of context on my part

cc @ruflin @webmat @clintongormley @jasontedor @bytebilly 

(first Elasticsearch issue! 🎉 )

### PR DESCRIPTION
**Pull request by jasontedor**
This commit enhances the required pipeline functionality by changing it so that default/request pipelines can also be executed, but the required pipeline is always executed last. This gives users the flexibility to execute their own indexing pipelines, but also ensure that any required pipelines are also executed. Since such pipelines are executed last, we change the name of required pipelines to final pipelines.

Closes #49247

### CODE REVIEW

File: qa/a.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+This commit adds coverage to the docs for some missing built-in shard
```
Comment (martijnvg) [inline qa/a.txt]: this seems unrelated to the change?

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 2]
```diff
@@ -98,6 +98,7 @@
     private XContentType contentType;
 
     private String pipeline;
+    private String finalPipeline;
 
     private boolean isPipelineResolved;
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: Is the `isPipelineResolved` field still needed and the corresponding if statement in `TransportBulkAction`?
Because before this change, there was validation whether a pipeline on a request was overwriting a required pipeline or in the case that there is both a default and required pipeline. This no longer seems to be the case, and therefor whether a pipeline has been resolved on coordinating node is not really important anymore?

**Review by martijnvg — COMMENTED** [review]
This looks good (besides the test failures). I left two comment.

File: qa/a.txt (hunk 1 above)
Comment (jasontedor) [reply qa/a.txt]: Oops, I removed this.

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java (hunk 2 above)
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: I can look into this in a follow-up. My aim is to get the final pipeline behavior in with the minimal change set since this issue is marked as a blocker for an upcoming release.

**Comment by jasontedor** [comment]
@elasticmachine update branch

**Review by martijnvg — APPROVED** [review]
LGTM
